src/audio/detect_test.c

```diff
@@ -0,0 +1,664 @@
+		diff = abs(*src) - cd->activation;
+		step = diff >> cd->config.activation_shift;
+
+		/* prevent taking 0 steps when the diff is too low */
+		cd->activation += !step ? diff : step;
+
+		if (cd->detect_preamble >= cd->keyphrase_samples) {
+			if (cd->activation >= cd->config.activation_threshold) {
+				detect_test_notify(dev);
+				cd->detected = 1;
+			}
+		} else {
+			++cd->detect_preamble;
+		}
+	}
+}
+
+static void free_mem_load(struct comp_data *cd)
+{
+	if (!cd) {
+		trace_keyword_error("free_mem_load() error: invalid cd");
+		return;
+	}
+
+	if (cd->load_memory) {
+		rfree(cd->load_memory);
+		cd->load_memory = NULL;
+	}
+}
+
+static int alloc_mem_load(struct comp_data *cd, uint32_t size)
+{
+	if (!cd) {
+		trace_keyword_error("alloc_mem_load() error: invalid cd");
+		return -EINVAL;
+	}
+
+	free_mem_load(cd);
+
+	if (!size)
+		return 0;
+
+	cd->load_memory = rballoc(RZONE_BUFFER, SOF_MEM_CAPS_RAM, size);
+
+	if (!cd->load_memory) {
+		trace_keyword_error("alloc_mem_load() alloc failed");
+		return -ENOMEM;
+	}
+
+	bzero(cd->load_memory, size);
+	cd->buf_copy_pos = 0;
+
+	return 0;
+}
+
+static void test_keyword_set_default_config(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	memset(&cd->config, 0, sizeof(cd->config));
+
+	cd->keyphrase_samples = KEYPHRASE_DEFAULT_PREAMBLE_LENGTH;
+	cd->config.activation_shift = ACTIVATION_DEFAULT_SHIFT;
+	cd->config.activation_threshold = ACTIVATION_DEFAULT_THRESHOLD_S16;
+}
+
+static struct comp_dev *test_keyword_new(struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *keyword;
+	struct sof_ipc_comp_process *ipc_keyword =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_data *cd;
+
+	trace_keyword("test_keyword_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_keyword->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_KEYWORD, ipc_keyword->config);
+		return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		      COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	keyword = (struct sof_ipc_comp_process *)&dev->comp;
+	memcpy_s(keyword, sizeof(*keyword), ipc_keyword,
+		 sizeof(struct sof_ipc_comp_process));
+
+	cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	/* using default processing function */
+	cd->detect_func = default_detect_test;
+
+	test_keyword_set_default_config(dev);
+
+	comp_set_drvdata(dev, cd);
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void test_keyword_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("test_keyword_free()");
+
+	free_mem_load(cd);
+	rfree(cd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int test_keyword_params(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	/* TODO: remove in the future */
+	dev->params.channels = 1;
+
+	if (dev->params.channels != 1) {
```

Comment:
CodeCheck complains that you have unrechable code here

Reply:
@tlauda @mrajwa may I remove line 261 now?